server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1992,72 +1990,79 @@ protected final Releasable acquireEnsureOpenRef() {
-            throw new IllegalStateException(e);
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            logger.error("interrupted while draining operations on close");
-            throw new IllegalStateException(e);
-        }
+        drainOnCloseListener.addListener(listener.map(unused -> true));
     }
 
     /**
      * Flush the engine (committing segments to disk and truncating the translog) and close it.
      */
-    public void flushAndClose() throws IOException {
+    public void flushAndClose(ActionListener<Void> listener) throws IOException {
         logger.trace("flushAndClose() maybe draining ops");
-        if (isClosed.get() == false && drainForClose()) {
-            logger.trace("flushAndClose drained ops");
-            try {
-                logger.debug("flushing shard on close - this might take some time to sync files to disk");
-                try {
-                    // TODO: We are not waiting for full durability here atm because we are on the cluster state update thread
-                    flushHoldingLock(false, false, ActionListener.noop());
-                } catch (AlreadyClosedException ex) {
-                    logger.debug("engine already closed - skipping flushAndClose");
+        var awaitPendingCloseListener = ActionListener.runBefore(listener, this::awaitPendingClose);
+        if (isClosed.get() == false) {
+            drainForClose(awaitPendingCloseListener.map(drained -> {
+                if (drained) {
+                    logger.trace("flushAndClose drained ops");
+                    try {
+                        logger.debug("flushing shard on close - this might take some time to sync files to disk");
+                        try {
+                            // TODO: We are not waiting for full durability here atm because we are on the cluster state update thread
+                            flushHoldingLock(false, false, ActionListener.noop());
+                        } catch (AlreadyClosedException ex) {
+                            logger.debug("engine already closed - skipping flushAndClose");
+                        }
+                    } finally {
+                        closeNoLock("flushAndClose", closedLatch);
+                    }
                 }
-            } finally {
-                closeNoLock("flushAndClose", closedLatch);
-            }
+                return null;
+            }));
+        } else {
+            awaitPendingCloseListener.onResponse(null);
         }
-        awaitPendingClose();
     }
 
     @Override
     public void close() throws IOException {
+        // TODO Called indirectly via try-with-resources or `IOUtils.close` from tests, Ideally, we would remove this method,
+        // and replace it with a blocking `EngineTestCase.close()` call, but it's used in a ton of tests and require
+        // refactoring, so we just check that the method is called from a test thread
```

Comment:
I'd rather we did this work up-front first to remove the `Closeable` interface from `Engine`. It may be rather noisy but it shouldn't be particularly complicated.

Reply:
Yeah, I tried to do that, there were a lot of changes required, especially in `InternalEngineTests` which is 7800 lines of code. Let me spin up a separate PR in trying to remove `Closeable` from `Engine`, maybe I will have some success with automatic refactoring tools :+1:

Reply:
I've submitted #111094